### PR TITLE
[FLINK-16866] Make job submission non-blocking

### PR DESCRIPTION
**Pull request by rmetzger**
## What is the purpose of the change

This is changing the semantics of the job submission: Instead of completing the `Dispatcher.submitJob()` future after all the initialization happened (which can potentially involve calling external systems etc.), the `.submitJob()` call now returns as soon as the job has been accepted by the Dispatcher.
The benefit of this change is that the users will see the root cause of a submission timeout, instead of an akka.ask.timeout.

## Brief change log

- Introduce a `DispatcherJob` abstraction that manages the job in a new `INITIALIZING` state
- Change web frontend to cope with initializing jobs
- change clients to submit & poll

## Verifying this change

This PR introduces various new tests for verification.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no 
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes 🔥 
  - The S3 file system connector: no 

## Documentation

This change is transparent to the user and doesn't need a documentation update.

### CODE REVIEW

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @rmetzger. I went half way through and added some comments. The main comment would be that the `DispatcherJob` seems to have an inconsistent semantically contract. I think this could be cleaned up a bit and made more explicit.

File: flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java [hunk 1]
```diff
@@ -111,4 +122,41 @@ public static void executeProgram(
...
+
+	/**
+	 * This method blocks until the job status is not INITIALIZING anymore.
+	 * If the job is FAILED, it throws an CompletionException with the failure cause.
+	 * @param jobStatusSupplier supplier returning the job status.
+	 */
+	public static void waitUntilJobInitializationFinished(
+		SupplierWithException<JobStatus, Exception> jobStatusSupplier,
+		SupplierWithException<JobResult, Exception> jobResultSupplier
+		) throws CompletionException {
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java:134]: For some reason, I assumed that `CompletionExceptions` are the proper way of signaling an error in futures. However, this assumption seems to be false, any RuntimeException will be wrapped into a CompletionException by the future.

I see solutions 2 here
a) declare the method throws any Throwable (then I have to use `FutureUtils.uncheckedFunction` when I use it within futures, to wrap errors in RuntimeExceptions
b) wrap any Throwable in a `RuntimeException` and remove the declaration on the method.

Unless you tell suggest otherwise, I'll go with b) for now.

File: flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java [hunk 2]
```diff
@@ -111,4 +122,41 @@ public static void executeProgram(
...
+				Thread.sleep(waitStrategy.sleepTime(attempt++));
+				status = jobStatusSupplier.get();
+			}
+			if (status == JobStatus.FAILED) {
+				// note: we can not distinguish between initialization failures and failures once
+				// execution has started. Execution errors are potentially reported here.
+				JobResult result = jobResultSupplier.get();
+				Optional<SerializedThrowable> throwable = result.getSerializedThrowable();
+				if (throwable.isPresent()) {
+					throw new CompletionException(throwable.get().deserializeError(Thread.currentThread().getContextClassLoader()));
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java:150]: My assumption is yes, but I agree this could be fragile.

If we are passing the wrong classloader here, we'll throw the `SerializedThrowable`, which contains contains the exception message and stacktrace as strings as well (users will see the cause, but they can't check for instance types in the causes).

Getting the usercode classloader into this method should be doable if you think it is the right approach. I just didn't want to be too invasive with this change :)

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java [hunk 3]
```diff
@@ -350,71 +341,42 @@ private boolean isPartialResourceConfigured(JobGraph jobGraph) {
 		}, getRpcService().getExecutor());
 	}
 
-	private CompletableFuture<Void> persistAndRunJob(JobGraph jobGraph) throws Exception {
+	private Void persistAndRunJob(JobGraph jobGraph) throws Exception {
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:344]: I have this ugly `Void` and `return null` here so that I can call the method as a method reference. 
My next push will contain an attempt to make this cleaner (by introducing a `VoidFunctionWithException` interface)

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java [hunk 4]
```diff
@@ -427,21 +389,39 @@ private JobManagerRunner startJobManagerRunner(JobManagerRunner jobManagerRunner
...
+						jobManagerSharedServices,
+						new DefaultJobManagerJobMetricGroupFactory(jobManagerMetricGroup),
+						fatalErrorHandler);
+					runner.start();
+					return runner;
+				} catch (Exception e) {
+					throw new CompletionException(new JobExecutionException(jobGraph.getJobID(), "Could not instantiate JobManager.", e));
+				}
+			},
+			rpcService.getExecutor()); // do not use main thread executor. Otherwise, Dispatcher is blocked on JobManager creation
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:418]: Done --> https://issues.apache.org/jira/browse/FLINK-19037

File: flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java:134]: I'd say let's go with c) Throw a properly defined exception which signals what went wrong and let the caller be responsible for wrapping it into the right format. The latter should not be the responsibility of this method which already does enough.

File: flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java:150]: I think it would be the right approach, yes.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:344]: shouldn't this also work if the method has the signature: `void persistAndRunJob(...)`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:418]: Great, thanks.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:344]: Unfortunately not: https://stackoverflow.com/questions/28762715/cannot-convert-void-to-java-lang-void

[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:344]: Hmm, I think the problem is that you are using `persistAndRunJob` still as a function even though it does not return anything. Changing `waitForTerminatingJobManager` to take a `ThrowingConsumer` should solve the problem.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+	private final long initializationTimestamp;
+	private final JobID jobId;
+	private final String jobName;
+
+	private final Object lock = new Object();
+
+	// internal field to track job status during initialization. Is not updated anymore after
+	// job is initialized, cancelled or failed.
+	@GuardedBy("lock")
+	private JobStatus jobStatus = JobStatus.INITIALIZING;
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:65]: This is a very good idea! Makes the code more concise.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	@Override
+	public CompletableFuture<Void> closeAsync() {
+		jobManagerRunnerFuture.handle((runner, throwable) -> {
+			if (throwable == null) {
+				// init was successful: close jobManager runner.
+				CompletableFuture<Void> jobManagerRunnerClose = jobManagerRunnerFuture.thenCompose(
+					AutoCloseableAsync::closeAsync);
+				FutureUtils.forward(jobManagerRunnerClose, terminationFuture);
+			} else {
+				// initialization has failed. Termination complete.
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:234]: You are right. I will complete the terminationFuture exceptionally.
I probably thought that the exception is already forwarded to the job result future, and handled accordingly.
But it is more consistent with the other case (job is already running, need to close the job manager runner).)

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+			"JobMaster Gateway is not available during initialization");
+		try {
+			return jobManagerRunnerFuture.thenCompose(JobManagerRunner::getJobMasterGateway);
+		} catch (Throwable e) {
+			throw new IllegalStateException("JobMaster gateway is not available", e);
+		}
+	}
+
+	@Override
+	public CompletableFuture<Void> closeAsync() {
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:226]: Unless we add a new `JobStatus.SUSPENDING` or `JobStatus.TERMINATING`, I don't see a good way of reporting to the user that we are waiting for the initialization to be finished so that we can close the JobManagerRunner.

Afaik this PR is not changing the semantics of a job stuck INITIALIZING. Previously, a job would report `CREATED` or `RUNNING` until it reaches the final `SUSPENDED` state, if `Dispatcher.onStop` get's called.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+						// cancellation will eventually complete the jobResultFuture
+						jobResultFuture.whenComplete((archivedExecutionGraph, resultThrowable) -> {
+							synchronized (lock) {
+								if (resultThrowable == null) {
+									jobStatus = archivedExecutionGraph.getState();
+								} else {
+									jobStatus = JobStatus.FAILED;
+								}
+							}
+						});
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:130]: > What happens if I call dispatcherJob.requestJob after the job has been cancelled in state INITIALIZING? I would like to obtain the final CANCELED ArchivedExecutionGraph ideally.

Since we are forwarding the the result future from the runner to the `DispatcherJob`'s result future, you should get a proper `ArchivedExecutionGraph` with the correct final state.

(note: I have removed this entire `.whenComplete()` stage, as it is not necessary to track the FAILED or CANCELLED etc. states).

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+		}
+	}
+
+	public CompletableFuture<ArchivedExecutionGraph> requestJob(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.requestJob(timeout));
+			} else {
+				return CompletableFuture.supplyAsync(() -> ArchivedExecutionGraph.createFromInitializingJob(jobId, jobName, null, jobStatus, initializationTimestamp));
+			}
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:204]: I think it will return the right result, but the naming of methods / variables here is terrible: `jobStatus == JobStatus.RUNNING` as soon as the `jobManagerRunnerFuture` completed. The job is not necessarily in RUNNING state. It could be in `CREATED` or `FAILING` ... we would still do the correct thing in `DispatcherJob`, but it is not very obvious.

I will improve the tracking of the job status inside the `DispatcherJob`, in particular the names (this seems to be the biggest flaw of my proposed changes).

[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:204]: Well, you are right, the result will still be wrong. We need to delegate to the resultFuture when it's complete.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:344]: Ah, very nice. Thanks a lot!

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+				return CompletableFuture.supplyAsync(() -> {
+					synchronized (lock) {
+						return ArchivedExecutionGraph.createFromInitializingJob(
+							jobId,
+							jobName,
+							jobStatus.asJobStatus(),
+							null,
+							initializationTimestamp);
+					}
+				});
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:199]: Why are you using `supplyAsync` here? `createFromInitializingJob` should be non-blocking.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+
+						// cancel job
+						jobManagerRunner
+							.getJobMasterGateway()
+							.thenCompose(gw -> gw.cancel(RpcUtils.INF_TIMEOUT))
+							.whenComplete((ignored, cancelThrowable) -> {
+							if (cancelThrowable != null) {
+								log.warn("Cancellation of job {} failed", jobId, cancelThrowable);
+							}
+						});
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java]: You are right. it is more elegant to schedule the cancel immediately.

**Comment by rmetzger** [comment]
Thanks a lot for your extensive and detailed review. It was very helpful for me!

I have addressed most of you your comments. I extended the `DispatcherJobTest`. 
I'm in the process of testing if the refactoring to the client has introduced any new test failures, and I might clean up my new tests in `DispatcherJob` (I would thus recommend you to review them last).

I might push a small change to the client code later today to better distinguish between initialization and runtime failures, so that we can solve the unstable `FunctionITCase.testInvalidUseOfTableFunction()` test.

File: flink-clients/src/test/java/org/apache/flink/client/ClientUtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Test for the ClientUtils.
+ */
+public class ClientUtilsTest extends TestLogger {
+
+	@ClassRule
```
Comment (rmetzger) [inline flink-clients/src/test/java/org/apache/flink/client/ClientUtilsTest.java:81]: This whole setup procedure seems way too involved. I guess a MiniCluster would do the job as well? Please +1 if you agree :)

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for addressing my comments @rmetzger. The code looks now much cleaner :-) I had a couple of additional comments. Next I will continue with reviewing the `DispatcherJob` tests.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java [hunk 13]
```diff
@@ -544,10 +705,13 @@ public void testPersistedJobGraphWhenDispatcherIsShutDown() throws Exception {
...
 
 		dispatcher.start();
+		jobMasterLeaderElectionService.isLeader(UUID.randomUUID());
 
 		final DispatcherGateway dispatcherGateway = dispatcher.getSelfGateway(DispatcherGateway.class);
-		final CompletableFuture<Acknowledge> submissionFuture = dispatcherGateway.submitJob(jobGraph, TIMEOUT);
-		submissionFuture.get();
+		dispatcherGateway.submitJob(jobGraph, TIMEOUT).get();
+
+		CommonTestUtils.waitUntilJobManagerIsInitialized(() -> dispatcherGateway.requestJobStatus(jobGraph.getJobID(), TIMEOUT).get());
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java:713]: Why do we have to wait for this here?

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java [hunk 14]
```diff
@@ -561,10 +725,12 @@ public void testPersistedJobGraphWhenDispatcherIsShutDown() throws Exception {
 	@Test
 	public void testJobSuspensionWhenDispatcherIsTerminated() throws Exception {
 		dispatcher = createAndStartDispatcher(heartbeatServices, haServices, new ExpectedJobIdJobManagerRunnerFactory(TEST_JOB_ID, createdJobManagerRunnerLatch));
+		jobMasterLeaderElectionService.isLeader(UUID.randomUUID());
 
 		DispatcherGateway dispatcherGateway = dispatcher.getSelfGateway(DispatcherGateway.class);
 
 		dispatcherGateway.submitJob(jobGraph, TIMEOUT).get();
+		CommonTestUtils.waitUntilJobManagerIsInitialized(() -> dispatcherGateway.requestJobStatus(jobGraph.getJobID(), TIMEOUT).get());
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java:733]: Why do we have to wait on this here?

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java [hunk 15]
```diff
@@ -603,6 +769,9 @@ public void testOnRemovedJobGraphDoesNotCleanUpHAFiles() throws Exception {
 			.build();
 		dispatcher.start();
 
+		// we need to elect a jobmaster leader to be able to cancel the job on the JobMaster.
+		jobMasterLeaderElectionService.isLeader(UUID.randomUUID()).get();
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java:773]: Why is this needed? Why do we need to cancel the job here?

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java [hunk 16]
```diff
@@ -617,16 +786,23 @@ public void testOnRemovedJobGraphDoesNotCleanUpHAFiles() throws Exception {
...
 
 		@Override
 		public TestingJobManagerRunner createJobManagerRunner(JobGraph jobGraph, Configuration configuration, RpcService rpcService, HighAvailabilityServices highAvailabilityServices, HeartbeatServices heartbeatServices, JobManagerSharedServices jobManagerSharedServices, JobManagerJobMetricGroupFactory jobManagerJobMetricGroupFactory, FatalErrorHandler fatalErrorHandler) throws Exception {
 			jobManagerRunnerCreationLatch.run();
 
-			return super.createJobManagerRunner(jobGraph, configuration, rpcService, highAvailabilityServices, heartbeatServices, jobManagerSharedServices, jobManagerJobMetricGroupFactory, fatalErrorHandler);
+			TestingJobManagerRunner runner = super.createJobManagerRunner(jobGraph, configuration, rpcService, highAvailabilityServices, heartbeatServices, jobManagerSharedServices, jobManagerJobMetricGroupFactory, fatalErrorHandler);
+			if (jobMasterGateway != null) {
+				runner.getJobMasterGateway().complete(jobMasterGateway);
+			}
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java:804]: Let's not use this hack here. We don't know whether `getJobMasterGateway` will always return the same future.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java [hunk 17]
```diff
@@ -690,4 +866,66 @@ public JobManagerRunner createJobManagerRunner(
...
+			super.initializeOnMaster(loader);
+
+			while (true) {
+				synchronized (lock) {
+					if (!blocking) {
+						return;
+					}
+					lock.wait(10);
+				}
+			}
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java:887]: I'd suggest to use a `OneShotLatch` here.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/TestingDispatcher.java [hunk 18]
```diff
@@ -31,11 +31,11 @@
 /**
  * {@link Dispatcher} implementation used for testing purposes.
  */
-class TestingDispatcher extends Dispatcher {
+public class TestingDispatcher extends Dispatcher {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/TestingDispatcher.java:34]: Please revert after updating the `ClientUtilsTest`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunnerITCase.java [hunk 19]
```diff
@@ -134,14 +135,13 @@ public void leaderChange_afterJobSubmission_recoversSubmittedJob() throws Except
 			final UUID firstLeaderSessionId = UUID.randomUUID();
 
 			final DispatcherGateway firstDispatcherGateway = electLeaderAndRetrieveGateway(firstLeaderSessionId);
-
 			firstDispatcherGateway.submitJob(jobGraph, TIMEOUT).get();
+			CommonTestUtils.waitUntilJobManagerIsInitialized(() -> firstDispatcherGateway.requestJobStatus(jobGraph.getJobID(), TIMEOUT).get());
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunnerITCase.java:139]: Why is this needed?

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/ZooKeeperDefaultDispatcherRunnerTest.java [hunk 20]
```diff
@@ -166,21 +167,28 @@ public void testResourceCleanupUnderLeadershipChange() throws Exception {
...
 				// initial run
-				DispatcherGateway dispatcherGateway = grantLeadership(dispatcherLeaderElectionService);
+				final DispatcherGateway dispatcherGateway = grantLeadership(dispatcherLeaderElectionService);
 
 				LOG.info("Initial job submission {}.", jobGraph.getJobID());
 				dispatcherGateway.submitJob(jobGraph, TESTING_TIMEOUT).get();
 
+				// wait until job is running
+				CommonTestUtils.waitUntilCondition(() ->
+					dispatcherGateway.requestJobStatus(jobGraph.getJobID(), TESTING_TIMEOUT).get() == JobStatus.RUNNING, Deadline.fromNow(VERIFICATION_TIMEOUT), 20L);
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/ZooKeeperDefaultDispatcherRunnerTest.java:177]: Why is this required here?

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/BlobsCleanupITCase.java [hunk 21]
```diff
@@ -181,17 +186,21 @@ private void testBlobServerCleanup(final TestCase testCase) throws Exception {
 		}
 
 		final CompletableFuture<JobSubmissionResult> submissionFuture = miniCluster.submitJob(jobGraph);
+		final JobSubmissionResult jobSubmissionResult = submissionFuture.get();
+
+		// wait until job is submitted
+		CommonTestUtils.waitUntilCondition(() -> miniCluster.getJobStatus(jobGraph.getJobID()).get() != JobStatus.INITIALIZING,
+			Deadline.fromNow(Duration.of(10, ChronoUnit.SECONDS)), 20L);
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/BlobsCleanupITCase.java:193]: Why do we have to wait on this?

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/LeaderChangeClusterComponentsTest.java [hunk 22]
```diff
@@ -103,6 +104,8 @@ public void testReelectionOfDispatcher() throws Exception {
 
 		submissionFuture.get();
 
+		CommonTestUtils.waitUntilJobManagerIsInitialized(() -> miniCluster.getJobStatus(jobId).get());
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/LeaderChangeClusterComponentsTest.java:107]: Why is this required?

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/LeaderChangeClusterComponentsTest.java [hunk 23]
```diff
@@ -135,6 +138,8 @@ public void testReelectionOfJobMaster() throws Exception {
...
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/LeaderChangeClusterComponentsTest.java:141]: Same here?

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java [hunk 24]
```diff
@@ -132,6 +132,7 @@ public void testJobRecoveryWithFailingTaskExecutor() throws Exception {
 	private CompletableFuture<JobResult> submitJobAndWaitUntilRunning(JobGraph jobGraph) throws Exception {
 		miniCluster.submitJob(jobGraph).get();
 
+		CommonTestUtils.waitUntilJobManagerIsInitialized(() -> miniCluster.getJobStatus(jobGraph.getJobID()).get());
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java:135]: Why is this needed here?

File: flink-table/flink-table-planner-blink/src/test/resources/log4j2-test.properties [hunk 25]
```diff
@@ -18,7 +18,7 @@
 
 # Set root logger level to OFF to not flood build logs
 # set manually to INFO for debugging purposes
-rootLogger.level = OFF
+rootLogger.level = DEBUG
```
Comment (tillrohrmann) [inline flink-table/flink-table-planner-blink/src/test/resources/log4j2-test.properties:21]: Please revert

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
I've gone through the tests and added some more comments. I believe that we can remove some checks that the job is initialized.

File: flink-clients/src/test/java/org/apache/flink/client/ClientUtilsTest.java (hunk 12 above)
Comment (rmetzger) [reply flink-clients/src/test/java/org/apache/flink/client/ClientUtilsTest.java:81]: The reason why I'm using real components is that I want to ensure the right exception is thrown from the `Dispatcher`: If somebody stops throwing a `JobInitializationException` during initialization in the Dispatcher, the test for `waitUntilJobInitializationFinished` should fail.

Comment (tillrohrmann) [reply flink-clients/src/test/java/org/apache/flink/client/ClientUtilsTest.java:81]: This could also be tested in two steps: 1. test the `ClientUtils` in isolation. 2. test that the `Dispatcher` throws such an exception under the required conditions (w/o involving the `ClientUtils`).

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/LeaderChangeClusterComponentsTest.java (hunk 23 above)
Comment (rmetzger) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/LeaderChangeClusterComponentsTest.java:141]: The next statement in the test is to revoke the JobMasterLeadership.
This revocation is only possible if we have leadership, which depends on the job being initialized.

File: flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java [hunk 26]
```diff
@@ -111,4 +122,48 @@ public static void executeProgram(
...
+			throw initializationException;
+		} catch (Throwable throwable) {
+			ExceptionUtils.checkInterrupted(throwable);
+			throw new JobInitializationException(jobID, "Error while waiting for job to be initialized", throwable);
+		}
+	}
+
+	public static <T> void waitUntilJobInitializationFinished(JobID id, ClusterClient<T> client, ClassLoader userCodeClassloader) throws
+		JobInitializationException {
+		waitUntilJobInitializationFinished(id, () -> client.getJobStatus(id).get(), () -> client.requestJobResult(id).get(), userCodeClassloader);
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java:167]: As far as I understand, the per job clusters are submitting jobs through `AbstractJobClusterExecutor.execute()`. They never use the `waitUntilJobInitializationFinished` utility on that codepath.

**Comment by rmetzger** [comment]
Thanks a lot for this extensive review. I believe I have addressed all comments. I'm looking forward to more feedback.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java [hunk 27]
```diff
@@ -687,7 +840,39 @@ public JobManagerRunner createJobManagerRunner(
 				jobManagerSharedServices,
 				jobManagerJobMetricGroupFactory,
 				fatalErrorHandler);
+
+			return jobManagerRunner;
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java:844]: nit: this could be reverted.

File: flink-tests/src/test/java/org/apache/flink/test/example/failing/JobSubmissionFailsITCase.java [hunk 28]
```diff
@@ -80,19 +75,6 @@ private static JobGraph getWorkingJobGraph() {
...
-
-	public JobSubmissionFailsITCase(boolean detached) {
-		this.detached = detached;
-	}
-
-	@Parameterized.Parameters(name = "Detached mode = {0}")
-	public static Collection<Boolean[]> executionModes(){
-		return Arrays.asList(new Boolean[]{false},
-				new Boolean[]{true});
-	}
```
Comment (tillrohrmann) [inline flink-tests/src/test/java/org/apache/flink/test/example/failing/JobSubmissionFailsITCase.java:95]: Why can we remove the detached test case here? Aren't we losing test coverage by removing this code?

Comment (rmetzger) [reply flink-tests/src/test/java/org/apache/flink/test/example/failing/JobSubmissionFailsITCase.java:95]: We can not distinguish between detached and attached job submission anymore. All submissions are detached.

This is the old code:
```java
			if (detached) {
				client.submitJob(jobGraph).get();
			} else {
				submitJobAndWaitForResult(client, jobGraph, getClass().getClassLoader());
			}
// method for reference:
	public static void submitJobAndWaitForResult(ClusterClient<?> client, JobGraph jobGraph, ClassLoader classLoader) throws Exception {
		client.submitJob(jobGraph)
			.thenCompose(client::requestJobResult)
			.get()
			.toJobExecutionResult(classLoader);
	}
```
... and with the non-blocking job submission, we would basically do the same in both cases.

**Comment by rmetzger** [comment]
Thanks a lot for your feedback! I addressed all comments and rebased the code to the latest master to get rid of the conflicts.

File: flink-tests/src/test/java/org/apache/flink/test/example/failing/JobSubmissionFailsITCase.java (hunk 28 above)
Comment (tillrohrmann) [reply flink-tests/src/test/java/org/apache/flink/test/example/failing/JobSubmissionFailsITCase.java:95]: Thanks for the explanation.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java [hunk 29]
```diff
@@ -107,18 +106,13 @@ private DispatcherJob(
...
 					jobManagerRunner.getResultFuture().whenComplete((archivedExecutionGraph, resultThrowable) -> {
-						if (archivedExecutionGraph != null) {
-							jobResultFuture.complete(DispatcherJobResult.forSuccess(archivedExecutionGraph));
-						} else {
-							jobResultFuture.completeExceptionally(ExceptionUtils.stripCompletionException(resultThrowable));
-						}
+						FutureUtils.forward(
+							jobManagerRunner.getResultFuture().thenApply(DispatcherJobResult::forSuccess),
+							jobResultFuture);
 					});
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java]: This was again ill advice because by calling `thenApply` we potentially introduce a `CompletionException` which will make the `DispatcherResourceCleanupTest.testBlobServerCleanupWhenJobNotFinished` fail. Hence, I would suggest to revert this change. 

On a side note: What I meant was actually to replace the `whenComplete` call with

```
FutureUtils.forward(
	  jobManagerRunner.getResultFuture().thenApply(DispatcherJobResult::forSuccess),
	  jobResultFuture);
```

**Review by tillrohrmann — APPROVED** [review]
Thanks for addressing my comments @rmetzger. The changes look good to me. There is one comment left which addresses a failing test failure. Given that AZP passes +1 for merging. Well done!

**Comment by tillrohrmann** [comment]
When closing the corresponding ticket, please add a release note stating the difference in submission semantics. E.g. when submitting a job one cannot immediately trigger job operation (triggering of savepoint, etc.). This is maybe one follow up issue we might wanna think through and adjust if it should become a usability problem.
